File: dominoes.py
import random

dominoes = [[0, 0], [0, 1], [0, 2], [0, 3], [0, 4], [0, 5], [0, 6],
            [1, 1], [1, 2], [1, 3], [1, 4], [1, 5], [1, 6], [2, 2],
            [2, 3], [2, 4], [2, 5], [2, 6], [3, 3], [3, 4], [3, 5],
            [3, 6], [4, 4], [4, 5], [4, 6], [5, 5], [5, 6], [6, 6]]
stock = []
player = []
computer = []
snake = []
status = None

for i in range(len(dominoes)):
    if i < 14:
        stock.append(dominoes.pop(random.choice(range(len(dominoes)))))
    elif i % 2 == 0:
        player.append(dominoes.pop(random.choice(range(len(dominoes)))))
    else:
        computer.append(dominoes.pop(random.choice(range(len(dominoes)))))

for i in range(7):
    if [6-i, 6-i] not in stock:
        if [6-i, 6-i] in player:
            player.remove([6-i, 6-i])
            status = 'computer'
        elif [6-i, 6-i] in computer:
            computer.remove([6-i, 6-i])
            status = 'player'
        snake.append([6-i, 6-i])
        break
    else:
        if i == 6:
            print('Error, no doubles!')

while True:
    print('=' * 70)
    print(f'Stock size: {len(stock)}')
    print(f'Computer pieces: {len(computer)}')
    print()
    if len(snake) > 6:
        temp_snake = snake[:3] + snake[-3:]
    else:
        temp_snake = snake
    for i, k in enumerate(temp_snake):
        print(k, end='')
        if len(snake) > 6 and i == 2:
            print('...', end='')
    print()
    print()
    print('Your pieces:')
    for i in range(len(player)):
        print(f'{i+1}:{player[i]}')
    print()
    if not computer:
        print('Status: The game is over. The computer won!')
        break
    if not player:
        print('Status: The game is over. You won!')
        break
    if snake[0][0] == snake[-1][-1]:
        flat_snake = []
        for i, j in snake:
            flat_snake.append(i)
            flat_snake.append(j)
        if flat_snake.count(flat_snake[0]) == 8:
            print("Status: The game is over. It's a draw!")
            break
    if status == 'player':
        print("Status: It's your turn to make a move. Enter your command.")
        while True:
            move = input()
            try:
                if abs(int(move)) <= len(player):
                    if abs(int(move)) == 0:
                        if stock:
                            player.append(stock.pop(random.choice(range(len(stock)))))
                    elif move.startswith('-'):
                        snake.insert(0, player.pop(abs(int(move))-1))
                    else:
                        snake.append(player.pop(abs(int(move))-1))
                    status = 'computer'
                    break
                else:
                    print('Invalid input. Please try again.')
            except ValueError:
                print('Invalid input. Please try again.')
    elif status == 'computer':
        input('Status: Computer is about to make a move. Press Enter to continue...')
        move = random.randint(-len(computer), len(computer))
        if move == 0:
            if stock:
                computer.append(stock.pop(random.choice(range(len(stock)))))
        elif move < 0:
            snake.insert(0, computer.pop(abs(move)-1))
        else:
            snake.append(computer.pop(abs(move)-1))
        status = 'player'

File: Dominoes.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class Dominoes {
    public static void main(String[] args) {
        List<List<Integer>> dominoes = new ArrayList<>(Arrays.asList(
                Arrays.asList(0, 0), Arrays.asList(0, 1), Arrays.asList(0, 2), Arrays.asList(0, 3),
                Arrays.asList(0, 4), Arrays.asList(0, 5), Arrays.asList(0, 6), Arrays.asList(1, 1),
                Arrays.asList(1, 2), Arrays.asList(1, 3), Arrays.asList(1, 4), Arrays.asList(1, 5),
                Arrays.asList(1, 6), Arrays.asList(2, 2), Arrays.asList(2, 3), Arrays.asList(2, 4),
                Arrays.asList(2, 5), Arrays.asList(2, 6), Arrays.asList(3, 3), Arrays.asList(3, 4),
                Arrays.asList(3, 5), Arrays.asList(3, 6), Arrays.asList(4, 4), Arrays.asList(4, 5),
                Arrays.asList(4, 6), Arrays.asList(5, 5), Arrays.asList(5, 6), Arrays.asList(6, 6)));

        List<List<Integer>> stock = new ArrayList<>();
        List<List<Integer>> player = new ArrayList<>();
        List<List<Integer>> computer = new ArrayList<>();
        List<List<Integer>> snake = new ArrayList<>();
        String status = null;
        Random random = new Random();

        for (int i = 0; i < dominoes.size(); i++) {
            if (i < 14) {
                stock.add(dominoes.remove(random.nextInt(dominoes.size())));
            } else if (i % 2 == 0) {
                player.add(dominoes.remove(random.nextInt(dominoes.size())));
            } else {
                computer.add(dominoes.remove(random.nextInt(dominoes.size())));
            }
        }

        for (int i = 0; i < 7; i++) {
            if (!containsDomino(stock, Arrays.asList(6 - i, 6 - i))) {
                if (containsDomino(player, Arrays.asList(6 - i, 6 - i))) {
                    player.remove(Arrays.asList(6 - i, 6 - i));
                    status = "computer";
                } else if (containsDomino(computer, Arrays.asList(6 - i, 6 - i))) {
                    computer.remove(Arrays.asList(6 - i, 6 - i));
                    status = "player";
                }
                snake.add(Arrays.asList(6 - i, 6 - i));
                break;
            } else {
                if (i == 6) {
                    System.out.println("Error, no doubles!");
                }
            }
        }

        while (true) {
            System.out.println("=".repeat(70));
            System.out.println("Stock size: " + stock.size());
            System.out.println("Computer pieces: " + computer.size());
            System.out.println();

            List<List<Integer>> tempSnake = snake.size() > 6 ? snake.subList(0, 3) : snake;
            tempSnake.addAll(snake.size() > 6 ? snake.subList(snake.size() - 3, snake.size()) : new ArrayList<>(snake));

            for (List<Integer> k : tempSnake) {
                System.out.print(k);
                if (snake.size() > 6 && tempSnake.indexOf(k) == 2) {
                    System.out.print("...");
                }
            }
            System.out.println("\n");

            System.out.println("Your pieces:");
            for (int i = 0; i < player.size(); i++) {
                System.out.println(i + 1 + ":" + player.get(i));
            }
            System.out.println();

            if (computer.isEmpty()) {
                System.out.println("Status: The game is over. The computer won!");
                break;
            }
            if (player.isEmpty()) {
                System.out.println("Status: The game is over. You won!");
                break;
            }

            if (snake.get(0).get(0).equals(snake.get(snake.size() - 1).get(1))) {
                List<Integer> flatSnake = new ArrayList<>();
                for (List<Integer> s : snake) {
                    flatSnake.addAll(s);
                }
                int first = flatSnake.get(0);
                if (flatSnake.stream().allMatch(x -> x == first)) {
                    System.out.println("Status: The game is over. It's a draw!");
                    break;
                }
            }

            if ("player".equals(status)) {
                System.out.println("Status: It's your turn to make a move. Enter your command.");
                while (true) {
                    String move = System.console().readLine();
                    try {
                        int index = Integer.parseInt(move);
                        if (Math.abs(index) <= player.size()) {
                            if (index == 0) {
                                if (!stock.isEmpty()) {
                                    player.add(stock.remove(random.nextInt(stock.size())));
                                }
                            } else if (index < 0) {
                                if (player.contains(player.get(Math.abs(index) - 1))) {
                                    List<Integer> domino = player.get(Math.abs(index) - 1);
                                    if (domino.get(1).equals(snake.get(0).get(0))) {
                                        snake.add(0, domino);
                                    } else if (domino.get(0).equals(snake.get(0).get(0))) {
                                        snake.add(0, Arrays.asList(domino.get(1), domino.get(0)));
                                    } else {
                                        System.out.println("Illegal move. Please try again.");
                                        continue;
                                    }
                                    player.remove(domino);
                                } else {
                                    System.out.println("Illegal move. Please try again.");
                                    continue;
                                }
                            } else {
                                if (player.contains(player.get(index - 1))) {
                                    List<Integer> domino = player.get(index - 1);
                                    if (domino.get(0).equals(snake.get(snake.size() - 1).get(1))) {
                                        snake.add(domino);
                                    } else if (domino.get(1).equals(snake.get(snake.size() - 1).get(1))) {
                                        snake.add(Arrays.asList(domino.get(1), domino.get(0)));
                                    } else {
                                        System.out.println("Illegal move. Please try again.");
                                        continue;
                                    }
                                    player.remove(domino);
                                } else {
                                    System.out.println("Illegal move. Please try again.");
                                    continue;
                                }
                            }
                            status = "computer";
                            break;
                        } else {
                            System.out.println("Invalid input. Please try again.");
                        }
                    } catch (NumberFormatException e) {
                        System.out.println("Invalid input. Please try again.");
                    }
                }
            } else if ("computer".equals(status)) {
                System.out.println("Status: Computer is about to make a move. Press Enter to continue...");
                while (true) {
                    int index = random.nextInt(computer.size() + 1) * (random.nextBoolean() ? 1 : -1);
                    if (index == 0) {
                        if (!stock.isEmpty()) {
                            computer.add(stock.remove(random.nextInt(stock.size())));
                        }
                    } else if (index < 0) {
                        if (computer.contains(computer.get(Math.abs(index) - 1))) {
                            List<Integer> domino = computer.get(Math.abs(index) - 1);
                            if (domino.get(1).equals(snake.get(0).get(0))) {
                                snake.add(0, domino);
                            } else if (domino.get(0).equals(snake.get(0).get(0))) {
                                snake.add(0, Arrays.asList(domino.get(1), domino.get(0)));
                            } else {
                                continue;
                            }
                            computer.remove(domino);
                        } else {
                            continue;
                        }
                    } else {
                        if (computer.contains(computer.get(index - 1))) {
                            List<Integer> domino = computer.get(index - 1);
                            if (domino.get(0).equals(snake.get(snake.size() - 1).get(1))) {
                                snake.add(domino);
                            } else if (domino.get(1).equals(snake.get(snake.size() - 1).get(1))) {
                                snake.add(Arrays.asList(domino.get(1), domino.get(0)));
                            } else {
                                continue;
                            }
                            computer.remove(domino);
                        } else {
                            continue;
                        }
                    }
                    status = "player";
                    break;
                }
            }
        }
    }

    private static boolean containsDomino(List<List<Integer>> list, List<Integer> target) {
        for (List<Integer> domino : list) {
            if (domino.equals(target)) {
                return true;
            }
        }
        return false;
    }
}
